internal/controllers/job_event_reconciler.go

```diff
@@ -0,0 +1,232 @@
+	"github.com/kubernetes-sigs/kernel-module-management/internal/meta"
+	"golang.org/x/exp/maps"
+	"golang.org/x/text/cases"
+	"golang.org/x/text/language"
+	v1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	JobEventReconcilerName = "JobEvent"
+
+	createdAnnotationKey = "kmm.node.kubernetes.io/created-event-sent"
+)
+
+type jobEvent struct {
+	jobType string
+}
+
+func (b *jobEvent) ReasonCreated() string {
+	return b.jobType + "Created"
+}
+
+func (b *jobEvent) ReasonFailed() string {
+	return b.jobType + "Failed"
+}
+
+func (b *jobEvent) ReasonSucceeded() string {
+	return b.jobType + "Succeeded"
+}
+
+func (b *jobEvent) String() string {
+	return b.jobType
+}
+
+var caser = cases.Title(language.English)
+
+func newJobEvent(eventType string) *jobEvent {
+	return &jobEvent{
+		jobType: caser.String(eventType),
+	}
+}
+
+type JobEventReconciler struct {
+	client   client.Client
+	recorder record.EventRecorder
+}
+
+func NewJobEventReconciler(client client.Client, eventRecorder record.EventRecorder) *JobEventReconciler {
+	return &JobEventReconciler{
+		client:   client,
+		recorder: eventRecorder,
+	}
+}
+
+func (r *JobEventReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	logger := ctrl.LoggerFrom(ctx)
+
+	pod := v1.Pod{}
+
+	if err := r.client.Get(ctx, req.NamespacedName, &pod); err != nil {
+		if k8serrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+
+		return reconcile.Result{}, fmt.Errorf("could not get pod %s: %v", req.NamespacedName, err)
+	}
+
+	jobType := pod.Labels[constants.PodType]
+
+	if jobType == "" {
+		return ctrl.Result{}, fmt.Errorf("job type not defined for Pod %s", req.NamespacedName)
+	}
+
+	je := newJobEvent(jobType)
+
+	kernelVersion := pod.Labels[constants.TargetKernelTarget]
+
+	mod := kmmv1beta1.Module{}
+
+	moduleNSN := types.NamespacedName{
+		Namespace: req.Namespace,
+		Name:      pod.Labels[constants.ModuleNameLabel],
+	}
+
+	if err := r.client.Get(ctx, moduleNSN, &mod); err != nil {
+		if k8serrors.IsNotFound(err) {
+			logger.Info(
+				"Module not found for job",
+				"pod name",
+				req.NamespacedName,
+				"module name",
+				moduleNSN,
+			)
+
+			return ctrl.Result{}, r.removeFinalizer(ctx, &pod)
+		} else {
+			return ctrl.Result{}, fmt.Errorf("error while getting Module %s: %v", moduleNSN, err)
+		}
+	}
+
+	eventAnnotations := map[string]string{
+		"kernel-version": kernelVersion,
+		"pod-name":       pod.Name,
+	}
+
+	if _, ok := pod.GetAnnotations()[createdAnnotationKey]; !ok {
```

Comment:
Why do we need this whole annotation handling? Can't we just create and event on PodRunning? Do we even want an event other than Failed or Succeeded for build/sign?

Reply:
The annotation is there to make sure that we do not send the `created` event twice.

Reply:
If the created event will be send on PodRunningPhase, won't it be sufficient? Reconciliation will be re-scheduled only in case something has changed in Pod status, which means that the Phase has changed from Running to Success/Failure, no?

Reply:
Let us consider the following scenario:
- pod P is running, we send the `created` event;
- our controller goes down and back up;
- when it starts, it processes all watched pods, including P

In that case we have no way to determine if we sent the event already or not.